src/aap_eda/services/activation/engine/common.py

```diff
@@ -113,6 +121,162 @@ class ContainerRequest(BaseModel):
     extra_args: tp.Optional[dict] = None
 
 
+class ContainerableMixinError(Exception):
+    """Base class for exceptions from implementers of ContainerableMixin."""
+
+    pass
+
+
+class ContainerableInvalidError(ContainerableMixinError):
+    pass
+
+
+class ContainerableNoLatestInstanceError(ContainerableMixinError):
+    pass
+
+
+# To use ContainerableMixin the model class adding the mixin is required to
+# have the following attributes (or property getters):
+#
+#   Attribute               Type
+#   ---------               ----
+#   decision_environment    DecisionEnvironment
+#   extra_var               ExtraVar
+#   latest_instance         RulebookProcess
+#   restart_policy          str
+#   rulebook_rulesets       str
+#
+class ContainerableMixin:
+    def get_command_line_parameters(self) -> dict[str, tp.Any]:
+        """Return parameters for running ansible-rulebook."""
+        self.validate()
+
+        access_token, refresh_token = create_jwt_token()
+        return {
+            "id": str(self.latest_instance.id),
+            "ws_url": self._get_ws_url(),
+            "log_level": settings.ANSIBLE_RULEBOOK_LOG_LEVEL,
+            "ws_ssl_verify": settings.WEBSOCKET_SSL_VERIFY,
+            "ws_token_url": self._get_ws_token_url(),
+            "ws_access_token": access_token,
+            "ws_refresh_token": refresh_token,
+            "heartbeat": settings.RULEBOOK_LIVENESS_CHECK_SECONDS,
+            "skip_audit_events": False,
+        }
+
+    def get_container_parameters(self) -> dict[str, tp.Any]:
+        """Return parameters used to create a ContainerRquest."""
+        self.validate()
+
+        return {
+            "credential": self._get_image_credential(),
+            "name": self._get_container_name(),
+            "image_url": self.decision_environment.image_url,
+            "ports": self._get_ports(),
+            "env_vars": settings.PODMAN_ENV_VARS,
+            "extra_args": settings.PODMAN_EXTRA_ARGS,
+            "mem_limit": settings.PODMAN_MEM_LIMIT,
+            "mounts": settings.PODMAN_MOUNTS,
+            "process_parent_id": self.id,
+            "rulebook_process_id": self.latest_instance.id,
+            "cmdline": self._build_cmdline(),
+        }
+
+    def get_container_request(self) -> ContainerRequest:
+        """Return ContainerRequest used for creation."""
+        params = self.get_container_parameters()
```

Comment:
What is the purpose of a method to build a dict to later call it here and compose the ContainerRequest from that dict when you could just build the ContainerRequest, directly? It looks like a redundant step that also impact to tests. Same for _build_cmdline function.

Reply:
AAP-20044 specifies the get* methods as part of the interface. Once those exist they act as the single source of truth for the respective values.

Reply:
Is maybe the ticket outdated/incorrect? @mkanoor I think doesn't make sense to have a get_container_parameters with a dict when all is contained in the dataclass ContainerRequest. That dataclass is the interface for their consumers, otherwise we would have it duplicated.  I would say only the ContainerRequest is needed. Problably not even we need to get the CmdLine alone as it is part of the ContainerRequest itself.

Reply:
Oh. That was your point? The way it was worded came across, to me, as a critique of not just directly accessing the attributes when creating a ContainerRequest (which would be non-stellar software engineering). 

What you explained here makes more sense.

Reply:
Resolving solely to get the code in. I’ll pursue this after returning from PTO. Feel free to discuss whether Alejandro’s suggestion meets everyone’s use cases and, if so, make the necessary changes, if desired.